pkg/storaged/client.js

```diff
@@ -206,6 +214,203 @@
+        return;
+    }
+
+    const uuids_subvols = { };
+    const uuids_usage = { };
+    const btrfs_default_subvol = { };
+    for (const uuid of Object.keys(client.uuids_btrfs_volume)) {
+        const blocks = client.uuids_btrfs_blocks[uuid];
+        if (!blocks)
+            continue;
+
+        // In multi device setups MountPoints can be on either of the block devices, so try them all.
+        const MountPoints = blocks.map(block => {
+            return client.blocks_fsys[block.path];
+        }).map(block_fsys => block_fsys.MountPoints).reduce((accum, current) => accum.concat(current));
+        const mp = MountPoints[0];
+        if (mp) {
+            let output;
+            const mount_point = utils.decode_filename(mp);
+            try {
+                // HACK: UDisks GetSubvolumes method uses `subvolume list -p` which
+                // does not show the full subvolume path which we want to show in the UI
+                //
+                // $ btrfs subvolume list -p /run/butter
+                // ID 256 gen 7 parent 5 top level 5 path one
+                // ID 257 gen 7 parent 256 top level 256 path two
+                // ID 258 gen 7 parent 257 top level 257 path two/three/four
+                //
+                // $ btrfs subvolume list -ap /run/butter
+                // ID 256 gen 7 parent 5 top level 5 path <FS_TREE>/one
+                // ID 257 gen 7 parent 256 top level 256 path one/two
+                // ID 258 gen 7 parent 257 top level 257 path <FS_TREE>/one/two/three/four
+                output = await cockpit.spawn(["btrfs", "subvolume", "list", "-ap", mount_point], { superuser: true, err: "message" });
+            } catch (err) {
+                console.error(`unable to obtain subvolumes for mount point ${mount_point}`, err);
+                return;
+            }
+            const subvols = [{ pathname: "/", id: 5 }];
+            for (const line of output.split("\n")) {
+                const m = line.match(/ID (\d+).*parent (\d+).*path (<FS_TREE>\/)?(.*)/);
+                if (m)
+                    subvols.push({ pathname: m[4], id: Number(m[1]) });
+            }
+            uuids_subvols[uuid] = subvols;
+
+            // HACK: Obtain the default subvolume, required for mounts in which do not specify a subvol and subvolid.
+            // In the future can be obtained via UDisks, it requires the btrfs partition to be mounted somewhere.
+            // https://github.com/storaged-project/udisks/commit/b6966b7076cd837f9d307eef64beedf01bc863ae
+            try {
+                output = await cockpit.spawn(["btrfs", "subvolume", "get-default", mount_point], { superuser: true, err: "message" });
+                const id_match = output.match(/ID (\d+).*/);
+                if (id_match)
+                    btrfs_default_subvol[uuid] = Number(id_match[1]);
+            } catch (err) {
+                console.error(`unable to obtain default subvolume for mount point ${mount_point}`, err);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20240111-133844-24a81545-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)